src/core/transport/chttp2/writing.c

```diff
@@ -0,0 +1,215 @@
+         stream_global->outgoing_window > 0) {
+    stream_writing->id = stream_global->id;
+    window_delta = grpc_chttp2_preencode(
+        stream_global->outgoing_sopb->ops, &stream_global->outgoing_sopb->nops,
+        GPR_MIN(transport_global->outgoing_window,
+                stream_global->outgoing_window),
+        &stream_writing->sopb);
+    GRPC_CHTTP2_FLOWCTL_TRACE_TRANSPORT(
+        "write", transport_global, outgoing_window, -(gpr_int64)window_delta);
+    GRPC_CHTTP2_FLOWCTL_TRACE_STREAM("write", transport_global, stream_global,
+                                     outgoing_window, -(gpr_int64)window_delta);
+    transport_global->outgoing_window -= window_delta;
+    stream_global->outgoing_window -= window_delta;
+
+    if (stream_global->write_state == WRITE_STATE_QUEUED_CLOSE &&
+        stream_global->outgoing_sopb->nops == 0) {
+      if (!transport_global->is_client && !stream_global->read_closed) {
+        stream_writing->send_closed = SEND_CLOSED_WITH_RST_STREAM;
+      } else {
+        stream_writing->send_closed = SEND_CLOSED;
+      }
+    }
+    if (stream_writing->sopb.nops > 0 ||
+        stream_writing->send_closed != DONT_SEND_CLOSED) {
+      grpc_chttp2_list_add_writing_stream(transport_writing, stream_writing);
+    }
+
+    /* we should either exhaust window or have no ops left, but not both */
+    if (stream_global->outgoing_sopb->nops == 0) {
+      stream_global->outgoing_sopb = NULL;
+      grpc_chttp2_schedule_closure(transport_global,
+                                   stream_global->send_done_closure, 1);
+    } else if (stream_global->outgoing_window > 0) {
+      grpc_chttp2_list_add_writable_stream(transport_global, stream_global);
+    }
+  }
+
+  /* for each grpc_chttp2_stream that wants to update its window, add that
+   * window here */
+  while (grpc_chttp2_list_pop_writable_window_update_stream(transport_global,
+                                                            &stream_global)) {
+    window_delta =
+        transport_global->settings[LOCAL_SETTINGS]
+                                  [GRPC_CHTTP2_SETTINGS_INITIAL_WINDOW_SIZE] -
+        stream_global->incoming_window;
+    if (!stream_global->read_closed && window_delta > 0) {
+      gpr_slice_buffer_add(
+          &transport_writing->outbuf,
+          grpc_chttp2_window_update_create(stream_global->id, window_delta));
+      GRPC_CHTTP2_FLOWCTL_TRACE_STREAM("write", transport_global, stream_global,
+                                       incoming_window, window_delta);
+      stream_global->incoming_window += window_delta;
+      grpc_chttp2_list_add_incoming_window_updated(transport_global,
+                                                   stream_global);
+    }
+  }
+
+  /* if the grpc_chttp2_transport is ready to send a window update, do so here
+   * also */
+  if (transport_global->incoming_window <
+      transport_global->connection_window_target * 3 / 4) {
```

Comment:
added a comment... I actually have a PR that changes this algorithm, so I'm going to leave this at a comment update for now.